### ISSUE TITLE
Some HTML-formatted text is completely removed from JSDoc Intellisense

### ISSUE THREAD

**Issue by thw0rted** [microsoft/vscode#110945]
- VSCode Version: 1.51.1
- OS Version: Win10

Steps to Reproduce:

1. Annotate something (in TS or JS) with a JSDoc comment, using `<pre>` or `<code>` tags for formatting
2. The content of the tag is erased in the Intellisense description of the thing

Does this issue occur when all extensions are disabled?: Yes

----

See [this Playground example](https://www.typescriptlang.org/play?#code/PQKhCgAIUhBSBGBDAzgSwMaTQOwC4CmATgGZIYGQDuaeAFpADwYD2AJgQHwAejwrHTpCQ42TAA5EuATz6SuAOighg4XIVLlKAMRYtIAbyiRuALkg4ArgFsExANzHp5hHoA2BEY4C+4cBww3JClIDzxIEnNdFntIYGBIABV3PDRxSCkkNhRIACJ4ZHQsdWIyCmpaBhExBVygA), where `A basic interface with <code>x</code> and <pre>y</pre>` renders as `A basic interface with and .`.

I'm aware that per https://github.com/microsoft/vscode/issues/31161 the official position is that HTML markup is "not supported", so by all means strip out tags, but completely removing text content is still a bug.  We can't control how third-party code writes their documentation, so the language service needs to handle it in a sane way.

**Comment by trivikr** [comment]
The `<pre>` and `<code>` blocks are allowed under [CommonMark spec 0.29 4.5 Fenced code blocks](https://spec.commonmark.org/0.29/#fenced-code-blocks).

**Comment by thw0rted** [comment]
Per the other issue I linked in the OP, they "support Markdown without HTML tag support".  That's... a position to take, I guess -- if I had to write an IDE I'd certainly want to farm out Markdown parsing to one of the squillion existing libraries rather than write a partial implementation myself -- but at least they're clear about it.  My concern is not with getting the tags *supported*, it's that stripping the tags should not also remove the content inside the tags.

**Comment by trivikr** [comment]
Copying description from related issue which was closed as duplicate.

- VSCode Version: 1.52.1
- OS Version: Darwin x64 19.6.0

Steps to Reproduce:

1. Open the [example code](https://www.typescriptlang.org/play?#code/PQKhCgAIUgeAHAfAQUgIwIYGcCWBjSHAOwBcBTAJwDMM8zIB3HEgC0ioHsOA6WYJKCGDhi5arXoAxLpADeUdlwBckIgFcAtmkoBucAF9w4ADZkS7FdI47IwYJAAqXYyRzxIFMhgAmWSACJ-I1AIaEhUTFwCUUoaOkZmNkwKbkFhGPF4gCEMCjkFZJV1LV0DI1NzNBUcihs7R2dXd08fP38I7HxCUliJBNZ0XO4goA) in VSCode.
2. Hover over `Foo`. The Tooltip reads "".
3. Hover over `Bar`. The Tooltip reads "A basic interface with bar."

The doc comment for `Foo` is compliant under [CommonMark spec 0.29 4.6 HTML Blocks kind 6](https://spec.commonmark.org/0.29/#html-blocks)
* **Start condition**: line begins the string `<` followed by `p`, followed by the string `>`
* **End condition**: line is followed by a blank line

<details>
<summary>Example Code</summary>

```ts
/**
 * <p>A basic interface with foo.</p>
 */
interface Foo {
  foo: number;
}

let f: Foo; // Tooltip reads ""

/**
 * A basic interface with bar.
 */
interface Bar {
  bar: number;
}

let b: Bar; // Tooltip reads "A basic interface with bar."

```

</details>

<details>
<summary>Screenshots</summary>

![Screen Shot 2021-01-06 at 10 27 58 PM](https://user-images.githubusercontent.com/16024985/103859222-87ce4900-506e-11eb-81f4-5b2a66973ae5.png)

![Screen Shot 2021-01-06 at 10 28 03 PM](https://user-images.githubusercontent.com/16024985/103859229-8bfa6680-506e-11eb-92da-ba5241017005.png)

</details>


Does this issue occur when all extensions are disabled?: Yes

**Comment by trivikr** [comment]
Some findings which could be helpful:
* The markdownRenderer uses marked from `vs/base/common/marked/marked`: https://github.com/microsoft/vscode/blob/6260e655bc606e204ad51ce5b1cb34216cd36dd3/src/vs/base/browser/markdownRenderer.ts#L11
* The renderer is markdown parser from https://github.com/markedjs/marked https://github.com/microsoft/vscode/blob/6260e655bc606e204ad51ce5b1cb34216cd36dd3/src/vs/base/common/marked/marked.js#L1-L5

The [Marked Demo](https://marked.js.org/demo/?text=%3Cp%3EA%20basic%20interface%20with%20foo.%3C%2Fp%3E&options=%7B%0A%20%22baseUrl%22%3A%20null%2C%0A%20%22breaks%22%3A%20false%2C%0A%20%22gfm%22%3A%20true%2C%0A%20%22headerIds%22%3A%20true%2C%0A%20%22headerPrefix%22%3A%20%22%22%2C%0A%20%22highlight%22%3A%20null%2C%0A%20%22langPrefix%22%3A%20%22language-%22%2C%0A%20%22mangle%22%3A%20true%2C%0A%20%22pedantic%22%3A%20false%2C%0A%20%22sanitize%22%3A%20false%2C%0A%20%22sanitizer%22%3A%20null%2C%0A%20%22silent%22%3A%20false%2C%0A%20%22smartLists%22%3A%20false%2C%0A%20%22smartypants%22%3A%20false%2C%0A%20%22tokenizer%22%3A%20null%2C%0A%20%22walkTokens%22%3A%20null%2C%0A%20%22xhtml%22%3A%20false%0A%7D&version=master) returns data as per [CommonMark spec 0.29 4.6 HTML Blocks kind 6](https://spec.commonmark.org/0.29/#html-blocks) for provided example.

<details>
<summary>Screenshot</summary>

![Screen Shot 2021-01-07 at 8 33 50 AM](https://user-images.githubusercontent.com/16024985/103918058-2d130c80-50c3-11eb-84a9-014bd7582c0e.png)

</details>

The example is also given in README under CLI usage https://github.com/markedjs/marked#usage
The issue seems to be in VSCode repo.

**Comment by thw0rted** [comment]
Do you think the problem is [here](https://github.com/microsoft/vscode/blob/6260e655bc606e204ad51ce5b1cb34216cd36dd3/src/vs/base/browser/markdownRenderer.ts#L328)?  I just searched for "strip" on a hunch, and found that any time `marked` hits an HTML tag in that case, it's completely ignored.

**Comment by mjbvz** [comment]
As discussed in #31161, we strip html while rendering JSDocs. Here's the main place where this happens:

https://github.com/microsoft/vscode/blob/6260e655bc606e204ad51ce5b1cb34216cd36dd3/src/vs/base/browser/markdownRenderer.ts#L220

We need more community feedback before enabling a subset of safe html tags in our markdown strings

**Comment by thw0rted** [comment]
My suggestion earlier was that instead of whitelisting tags to let through, you could remove all markup and include only the text node(s) as bare strings. Of course for elements that change plain text layout (block tags like headings, paragraph, lists, etc) the result would look pretty ugly, but for inline tags it would be an improvement over current behavior without possibility of security ramifications. Of course a more thoughtful whitelisting would be more useful, but also more work.

**Comment by trivikr** [comment]
@thw0rted please use inclusive word like "allowlist"

**Comment by mjbvz** [comment]
Duplicate of #40607

**Comment by thw0rted** [comment]
@mjbvz I don't actually think this is a duplicate of the linked issue.  That issue wants (certain) HTML tags to be rendered into styled markup.  It's a pretty big bite to take, has a lot of security implications etc, and it looks to me like the issue hasn't seen movement in 3+ years as a result.

That's not what I'm asking for.  Today, right now, the text content of unsupported tags *disappears completely*, resulting in a useless mess for a lot of third party libraries that I do not control.  If the documentation says `/** Do <strong>not</strong> call the Foo method after this one */` it will render in the tooltip as `Do call the Foo method...`.  It's a bug with a simple, uncontroversial fix: stop eating the text.  I'm worried that closing this as a duplicate of a difficult, controversial issue will result in users being stuck with the current, wrong behavior until someone has the time and energy to make a bunch of security and design decisions.

**Comment by thw0rted** [comment]
Re-reading #40607, maybe @trivikr's issue is actually the same? There might actually be multiple separate but related disappearing-text issues.  I didn't initially understand the older issue just from reading through it, and I think it might only be the `code` and `pre` tags that have my specific problem.  I've tried a bunch of different tags in different places, and here's what I figured out:

* `code` and `pre` contents always disappear, at any position in the docs.  This is my problem.
*  `p` contents only disappear (per @trivikr's issue) when they begin at the start of a line.  The string `/** Some text <p>other text</p> */` renders as `Some text other text` but `/** <p>Some text</p> */` gives no tooltip.  Experimentally, any block-level tag has this bug, and any inline tag (except `code`!) does not.

Re-reading #40607, the OP wants the second issue resolved -- and also wants `<br/>` to translate to a forced newline (akin to ending Markdown with a double space), I guess?.

So, there are really 3 related issues.  `code` and `pre` are being mishandled regardless of position.  All block-level tags are being mishandled if they begin a new line.  And always-empty tags (`br` and `hr`) that should at least force a newline, don't.  If it makes sense to only fix one of those when all 3 can be fixed, I guess it makes sense to merge them, but I was hoping that it might be simpler to fix the `code`/`pre` issue.
